### PR TITLE
add nvidia/cutlass

### PR DESCRIPTION
**Pull request by h-vetinari**
Package https://github.com/NVIDIA/cutlass, with the goal of packaging https://github.com/openai/triton and eventually https://github.com/facebookresearch/xformers

### CODE REVIEW

**Comment by h-vetinari** [comment]
Ah, the tests seem to need a GPU, which the CI doesn't have...
```
*** Error: Could not detect active GPU device ID [no CUDA-capable device is detected]
```

**Comment by h-vetinari** [comment]
> I was trying to look for recipes to lint for you, but couldn't find any.
> Please ping the 'conda-forge/core' team (using the @ notation in a comment) if you believe this is a bug.

@conda-forge/core, there's a bug with the linter here. As of https://github.com/conda-forge/staged-recipes/pull/17303/commits/23186f95d5f7e804ae7550c300bb5ccc325d5dd2, there's definitely a recipe under `recipes/cutlass`, so something's off.

File: recipes/cutlass/build.sh
Comment (kkraus14) [inline recipes/cutlass/build.sh:12]: I believe this was fixed upstream in cmake 3.20: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5777

Comment (h-vetinari) [reply recipes/cutlass/build.sh:12]: Thanks! Will try removing it. I have amassed a "toolkit" of workarounds if things don't work out of the box (mostly through the help of the ever-patient @jaimergp), and that one was one of the defaults - would be very happy to hear if it's obsolete now. :)

**Comment by h-vetinari** [comment]
So, one thing is that things seem to get installed into `lib64` instead of `lib`, and I don't see where the switch should be in the upstream [`CMakeLists.txt`](https://github.com/NVIDIA/cutlass/blob/v2.8.0/CMakeLists.txt).

The other thing is that I saw some rpath-[manipulation](https://github.com/NVIDIA/cutlass/blob/v2.8.0/CMakeLists.txt#L518-L528) that we might want to patch out (esp the `$ORIGIN` ones)? Copied here for convenience. @conda-forge/help-c-cpp, any preferences or advice?
```
if(NOT WIN32)
  # Add common library search paths so executables and libraries can load and run
  # without LD_LIBRARY_PATH being set.
  link_libraries(
    "-Wl,-rpath,'$ORIGIN'"
    "-Wl,-rpath,'$ORIGIN/../lib64'"
    "-Wl,-rpath,'$ORIGIN/../lib'"
    "-Wl,-rpath,'${CUDA_TOOLKIT_ROOT_DIR}/lib64'"
    "-Wl,-rpath,'${CUDA_TOOLKIT_ROOT_DIR}/lib'"
    )
endif()
```

File: recipes/cutlass/build.sh
Comment (h-vetinari) [reply recipes/cutlass/build.sh]: Removed it, and it worked! :)

File: recipes/cutlass/build.sh
Comment (kkraus14) [inline recipes/cutlass/build.sh:11]: I believe all of this should already be taken care of by the compilers packages setting appropriate things

**Comment by h-vetinari** [comment]
OK, this now builds 🥳  (Magic variable turned out to be `CMAKE_LIBRARY_OUTPUT_DIRECTORY`)

I'm still a bit worried about the following though. Could someone either confirm or deny that we should patch this @conda-forge/cudatoolkit @jaimergp @xhochy?

> I saw some rpath-[manipulation](https://github.com/NVIDIA/cutlass/blob/v2.8.0/CMakeLists.txt#L518-L528) that we might want to patch out (esp the `$ORIGIN` ones)? Copied here for convenience.
> 
> ```
> if(NOT WIN32)
>   # Add common library search paths so executables and libraries can load and run
>   # without LD_LIBRARY_PATH being set.
>   link_libraries(
>     "-Wl,-rpath,'$ORIGIN'"
>     "-Wl,-rpath,'$ORIGIN/../lib64'"
>     "-Wl,-rpath,'$ORIGIN/../lib'"
>     "-Wl,-rpath,'${CUDA_TOOLKIT_ROOT_DIR}/lib64'"
>     "-Wl,-rpath,'${CUDA_TOOLKIT_ROOT_DIR}/lib'"
>     )
> endif()
> ```

**Comment by leofang** [comment]
I thought conda-build would prune unnecessary rpaths using patchelf/lief, so that we don't have to worry?

**Comment by h-vetinari** [comment]
> I thought conda-build would prune unnecessary rpaths using patchelf/lief, so that we don't have to worry?

Ah, that's amazing if correct. I wasn't aware of that.

**Comment by h-vetinari** [comment]
Well, assuming the rpaths set by the package get cleaned up correctly by patchelf/lief, then this is ready for review @conda-forge/staged-recipes

**Comment by h-vetinari** [comment]
Sorry, forgot to remove the "WIP: " from the title...

File: recipes/cutlass/conda_build_config.yaml
Comment (kkraus14) [inline recipes/cutlass/conda_build_config.yaml:5]: Can these be removed now that we can pin the compiler version per CUDA version?

**Comment by h-vetinari** [comment]
@kkraus14 
I took care of all the points you raised - very happy to see that the previous work-arounds are not necessary anymore. :)

**Review by kkraus14 — APPROVED** [review]

**Comment by h-vetinari** [comment]
Thanks @kkraus14! Anyone from @conda-forge/staged-recipes wants to have a look as well? 🙃

File: recipes/cutlass/LICENSE.txt
Comment (carterbox) [inline recipes/cutlass/LICENSE.txt:23]: OPTIONAL: You don't need to include the license as a file in the recipe if it is already included in the source tarball. The source is unpacked before the license file is extracted, so naming LICENSE.txt is sufficient.

File: recipes/cutlass/meta.yaml
Comment (carterbox) [inline recipes/cutlass/meta.yaml:13]: There is no `bld.bat`, but you aren't skipping Windows builds? Do you plan on building for Windows? Once this recipe is placed into a feedstock, Windows CUDA builds will become available, they just aren't used to validate proposed recipes.

**Review by carterbox — CHANGES_REQUESTED** [review]
Looks like the recipe [isn't building anything](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=456926&view=logs&j=232b7f5e-8d64-5438-fada-8478f379d226&t=e8516d7e-1153-5090-5807-53ae833fddd8&l=676). Not sure this is the fault of the recipe or not.

File: recipes/cutlass/build.sh
Comment (carterbox) [inline recipes/cutlass/build.sh:22]: [You are required](https://conda-forge.org/docs/maintainer/knowledge_base.html#cuda-builds) to add conda-forge's CMAKE_ARGS variable.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by carterbox** [comment]
@isuruf, As far as I can tell, all of the `linux_64_cuda_` builds have not actually compiled anything. Conda-build just says "Computed that there are 0 distributions to build from 1 recipes. Nothing to do", then exits. Any idea why?

**Comment by isuruf** [comment]
Needs https://github.com/conda-forge/staged-recipes/pull/16257/commits/8bbc3bebe6eeedbf4cd88fdea770d52427920e53

**Comment by h-vetinari** [comment]
> Needs [8bbc3be](https://github.com/conda-forge/staged-recipes/commit/8bbc3bebe6eeedbf4cd88fdea770d52427920e53)

Thanks. Cherry-picked from #16257

**Comment by h-vetinari** [comment]
Curiously, on every cuda version _but_ for 11.0, I get variations of:
```
lib/gcc/x86_64-conda-linux-gnu/9.4.0/include/avx512dqintrin.h(1372): error: identifier "__builtin_ia32_fpclasssd" is undefined
```
which surprises me, because the library version and the compiler are of course matching.

I also don't understand why it's cuda-version dependent...

Closest related thing I found is [this](https://stackoverflow.com/questions/60824242/suddenly-getting-builtin-ia32-sqrtsd-round-is-undefined-with-nvcc-gcc), but not sure how to apply that here.

**Comment by h-vetinari** [comment]
So CUDA 11.0 actually ran through 🥳 
Any tips on how to handle the suddenly-missing avx512 intrinsics would be much appreciated @kkraus14 @leofang 🙃

**Comment by leofang** [comment]
I have no idea... last time I hit an intrinsic-related failure, I ended up updating my compiler version, suspecting it's a compiler bug...

**Comment by h-vetinari** [comment]
@isuruf, could it be that other selectors are also affected by what https://github.com/conda-forge/staged-recipes/commit/8bbc3bebe6eeedbf4cd88fdea770d52427920e53 fixes?

Because the exploratory compiler bump in cbc.yaml here does _not_ get picked up with a `# [linux]` selector.

**Comment by isuruf** [comment]
Yes, variant config files have higher priority than the ones in `recipe/conda_build_config.yaml`. You'll have to change the type from variant config file to an exclusive config file. See https://github.com/conda/conda-build/blob/a3af116b7192a31d000e4b36e766f73368310901/conda_build/variants.py#L175-L181

**Comment by h-vetinari** [comment]
> Yes, variant config files have higher priority than the ones in `recipe/conda_build_config.yaml`.

OK, thank you. I was just surprised because cbc-in-staged-recipes used to work previously (e.g. for setting compiler versions), but IIUC the CUDA builds here now use variants that overrule this.

In any case, I don't want to change the CI setup here, so for now I'm just patching it in to see if it'd pass with GCC 10. That should obviously not be committed, but I'd easily redo it on the feedstock.

**Comment by h-vetinari** [comment]
> In any case, I don't want to change the CI setup here

I guess this would have been less ambiguous as "I'm not trying to change the CI setup", because I wouldn't want to presume I should, nor that that'd be welcome, nor mix that in with this PR.

> It would be better if you could help change the CI setup.

I hadn't heard about exclusive configs until today. Are you saying the cuda variants should all be made exclusive (so that recipes have a chance to override them with cbc again)?

TBH, I don't understand why anything overrides the recipe-cbc (the first and arguably only reasonable go-to for maintainers to change pinnings), but that's then perhaps a conda-build question.

**Comment by h-vetinari** [comment]
> > Are you saying the cuda variants should all be made exclusive (so that recipes have a chance to override them with cbc again)?
> 
> Yes.

OK, I'll try. This PR seems a good testing ground, because - apparently - things pass with GCC 10 and fail with GCC 9. So If I get it wrong (after readding cbc), I'll get a red CI pretty much right away. But assuming I get there, we can still break it out into a separate PR after.

> > TBH, I don't understand why anything overrides the recipe-cbc (the first and arguably only reasonable go-to for maintainers to change pinnings), but that's then perhaps a conda-build question.
> 
> Otherwise, it'd be impossible to select one variant. For eg: conda-smithy's generated variants in feedstocks will get less priority than the recipe/cbc which can have multiple variants.

That makes a lot of sense, thanks for the explanation.

**Comment by leofang** [comment]
Oh! So we hit another compiler bug?!

**Comment by h-vetinari** [comment]
> Oh! So we hit another compiler bug?!

Looks like it! But curiously, only on CUDA 11.1 & 11.2, not on CUDA 11.0 (ignore the current CI failure, this is running against an incompatible GCC-version, because I was too lazy to do all the cuda-version-specific cbc magic).

**Comment by h-vetinari** [comment]
@isuruf, https://github.com/conda-forge/staged-recipes/pull/17303/commits/fb3c9b66d020cb3addbfa3fe224b63266752b1d4 provides a working solution (recipe/cbc is respected). Would you like to review? Here or in a separate PR?

**Comment by h-vetinari** [comment]
> [fb3c9b6](https://github.com/conda-forge/staged-recipes/commit/fb3c9b66d020cb3addbfa3fe224b63266752b1d4) provides a working solution (recipe/cbc is respected). Would you like to review? Here or in a separate PR?

Opened #18024

**Comment by h-vetinari** [comment]
I tried copying the content of https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2010 into cbc.yaml, but it didn't work. Would be happy to hear how I can set GCC=10 for CUDA 11.1/11.2, but leave GCC 9 for CUDA 11.0.

**Comment by h-vetinari** [comment]
Any other difference between staged-recipes and a "normal" feedstock I'm overlooking here? From what I can tell, using parts of the pinning-cbc verbatim should always work (but doesn't here)?

**Comment by h-vetinari** [comment]
> > I tried copying the content of https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2010 into cbc.yaml, but it didn't work. Would be happy to hear how I can set GCC=10 for CUDA 11.1/11.2, but leave GCC 9 for CUDA 11.0.
>
> Any other difference between staged-recipes and a "normal" feedstock I'm overlooking here?

Could it be that it's due to staged-recipes not using conda-smithy, and hence not knowing about [`CF_CUDA_ENABLED`](https://github.com/conda-forge/conda-smithy/pull/1544)?

In that case, I won't get the recipe to pass here fully (but should be fine on the feedstock... 🤞) --- CUDA 11.0 needs GCC 9, and CUDA 11.1/11.2 (in this recipe) _needs_ GCC 10.

The only other way I see is to try patch the logic for CF_CUDA_ENABLED into [`build_all.py`](https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/build_all.py) somehow, if we're OK with the code duplication with conda-smithy... (we're talking about ~10 lines of code, under the assumption that setting the environment variable would be enough)

If I'm right with this summary, this would also mean that staged-recipes in its current state will not be able to reflect https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2010 once merged.

CC @conda-forge/core ([for](https://github.com/conda-forge/conda-smithy-feedstock/blob/master/recipe/meta.yaml#L61) conda-smithy)

**Comment by isuruf** [comment]
> If I'm right with this summary, this would also mean that staged-recipes in its current state will not be able to reflect https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2010 once merged.

Nope. The version is determined by the config files. For eg: https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64_cuda102.yaml#L7 which we can edit after conda-forge-pinning PR is merged.

**Comment by h-vetinari** [comment]
> Nope. The version is determined by the config files. For eg: https://github.com/conda-forge/staged-recipes/blob/main/.ci_support/linux64_cuda102.yaml#L7 which we can edit after conda-forge-pinning PR is merged.

Ah, makes sense, just that it's hard-coded and can't be overridden by cbc. Well, lucky that https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2010 seems to be nearing completion, otherwise this one would be stuck 😅

**Comment by h-vetinari** [comment]
Gentle ping @conda-forge/staged-recipes 🙃

File: recipes/cutlass/meta.yaml
Comment (carterbox) [inline recipes/cutlass/meta.yaml:15]:
```suggestion
  run_exports:
    - {{ pin_subpackage('cutlass', max_pin='x.x') }}

```
This package includes headers and a shared library, so it it meant to be linked against. Thus, it should have run_exports, so that it is automatically included as a dependency for packages that link to it. Please ensure that this version pinning is appropriate.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/cutlass/meta.yaml
Comment (h-vetinari) [reply recipes/cutlass/meta.yaml:15]: Thanks, I was considering this as well (but wanted to see if there's any other relevant feedback before retriggering CI).

**Comment by h-vetinari** [comment]
@kerrmudgeon
Just FYI, the packaging for conda-forge is almost there. 🙃

I wanted to ask if you could provide any information on the binary compatibility policies of cutlass? I.e. would a library built against 2.8.0 be able to load 2.9.0 at runtime?

CC @kkraus14 @jakirkham @leofang

**Review by carterbox — APPROVED** [review]

**Comment by carterbox** [comment]
I'm going to merge this PR. You can relax the run_export in your new feedstock.

**Comment by h-vetinari** [comment]
Thanks for review and merge @carterbox!

**Pull request merged**
